### PR TITLE
Add ready for use RRule instance with caching

### PR DESCRIPTION
**Pull request by polRk**
Add RRule ready for use instance with result caching

### Checklist

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by polRk** [comment]
I do not know how to cache `after` method without recurrence rule until value. Can u help me?

**Comment by polRk** [comment]
I'm inspired by this repository https://github.com/jakubroztocil/rrule

File: lib/src/rrule.dart [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+  @visibleForTesting
+  Cache get cache => _cache;
+
+  List<DateTime> before({
+    required DateTime date,
+    bool inc = false,
+  }) {
+    final instances = rule.getInstances(start: start);
+    final argumentsHash = hashList([instances, date, inc]);
```
Comment (polRk) [inline lib/src/rrule.dart:26]: Maybe it can be better?

```suggestion
    final argumentsHash = hashList([rule, start, date, inc]);
```

**Comment by JonasWanke** [comment]
Thanks for this contribution! Caching seems like a really useful feature as the internal calculations in this package can get quite complex.

I have a few questions/comments/suggestions about the architecture of this caching:

* Having a class called `RRule` (the common abbreviation for Recurrence Rule) next to the class `RecurrenceRule` seems a bit confusing. Could we either integrate this functionality directly in `RecurrenceRule` or maybe rename `RRule` to `CachingRecurrenceRule`?
* rrule.js's `before` and `after` only return a single value (the first event before/after the given instant) as opposed to all values before/after that instant, but I think that the latter is useful as well. (And if we need the former in the future, we can still add methods calculating and returning only the single next event.) However, I'm not sure whether we really need three separate methods (`before`, `after`, and `between`). Maybe we can combine those into a single method with optional `before` and `after` parameters?

  ```dart
  List<DateTime> getAllInstances({
    required DateTime start, // Only necessary if this is integrated into `RecurrenceRule` directly.
    DateTime? after,
    bool includeAfter = false,
    DateTime? before,
    bool includeBefore = false,
  }) {/* ... */}
  ```

* Currently, a hash of all relevant parameters (rrule, start, and method arguments) is used as the cache key. While hash functions try to minimize the risk of collisions, those are not impossible. Hence, I recommend creating a utility class like `_CacheKey` which stores those arguments directly and, as the name suggests, is used as the cache key. And as the cache is currently stored per recurrence rule, the rrule doesn't even have to be part of this key.
* `RRule` accepts an `initialCache`, similar to rrule.js. Do you know of any use-cases for supplying that initial cache state, especially since the `cache` itself is not made public (and hence can't be persisted otherwise by users of this package)?

**Comment by polRk** [comment]
> RRule accepts an initialCache, similar to rrule.js. Do you know of any use-cases for supplying that initial cache state, especially since the cache itself is not made public (and hence can't be persisted otherwise by users of this package)?

This can reduce the cold start. If I have 1000 recurring tasks, it will be better to use cached values rather than calculated ones.

File: lib/src/recurrence_rule.dart [hunk 2]
```diff
@@ -30,6 +31,7 @@ class RecurrenceRule {
     Set<int> byMonths = const {},
     Set<int> bySetPositions = const {},
     this.weekStart,
+    this.shouldCacheResults = false,
```
Comment (polRk) [inline lib/src/recurrence_rule.dart:34]: Protection against unexpected memory leaks

File: lib/src/recurrence_rule.dart [hunk 3]
```diff
@@ -121,38 +124,47 @@ class RecurrenceRule {
...
   bool get hasByWeeks => byWeeks.isNotEmpty;
 
   /// Corresponds to the `BYMONTH` property.
   final Set<int> byMonths;
+
   bool get hasByMonths => byMonths.isNotEmpty;
 
   /// Corresponds to the `BYSETPOS` property.
   final Set<int> bySetPositions;
+
```
Comment (JonasWanke) [inline lib/src/recurrence_rule.dart:167]: Please undo adding these empty lines

File: lib/src/cache.dart [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import 'package:meta/meta.dart';
+
+import 'utils.dart';
+
+class CacheKey {
+  CacheKey({
```
Comment (JonasWanke) [inline lib/src/cache.dart:6]:
```suggestion
@immutable
class CacheKey {
  const CacheKey({
```

File: lib/src/cache.dart [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+  DateTime start;
+
+  DateTime? after;
+
+  bool includeAfter;
+
+  DateTime? before;
+
+  bool includeBefore;
```
Comment (JonasWanke) [inline lib/src/cache.dart:22]:
```suggestion
  final DateTime start;
  final DateTime? after;
  final bool includeAfter;
  final DateTime? before;
  final bool includeBefore;
```

File: lib/src/cache.dart [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+class Cache {
+  final Map<CacheKey, List<DateTime>> _results = {};
+
+  @visibleForTesting
+  Map<CacheKey, List<DateTime>> get results => _results;
+
+  void add(CacheKey key, List<DateTime> data) {
+    _results.addAll({key: data});
```
Comment (JonasWanke) [inline lib/src/cache.dart:55]:
```suggestion
    _results[key]= data;
```

File: test/cache_test.dart [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  test('should add data to the cache', () {
+    final cache = Cache();
+    final key = CacheKey(start: DateTime.now());
+    final results = <DateTime>[];
+    cache.add(key, results);
+
+    expect(cache.results, containsPair(key, results));
+  });
+
+  test('should return date from the cache by key', () {
```
Comment (JonasWanke) [inline test/cache_test.dart:14]:
```suggestion
  test('should return data from the cache by key', () {
```

**Review by JonasWanke — CHANGES_REQUESTED** [review]
This looks great! I've added some minor comments, and then this can get merged :)

**Review by JonasWanke — APPROVED** [review]

**Comment by JonasWanke** [comment]
This is now published as part of [v0.2.3](https://pub.dev/packages/rrule/changelog#v023--2021-05-06). Thanks for contributing!

**Pull request merged**